core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
```

Comment:
> Good question. Well, you can't send the rocket. I guess it means that you don't know, and have to restart. I don't know if Kafka has timeouts for such commits, and if it would timeout that would only mean "you don't know if it was committed or not", i.e. you have to restart. In our case that means stream failure.

But this can be all handled inside the Source. I still don't see much value in exposing the CommitConfirmation since as a *user* I have no tool in my hand to redeliver the commit. I would rather have this behavior as configuration of the Source and only emitting elements after they have been commit-confirmed (retrying if necessary) and failing the stream if commit-confirmation did not arrive in time.

I guess the basic mismatch is how low-level the API should be. I personally prefer higher level tools since the client is always there for low level access.